File: iOS/Code/Swift/bugTrap/bugTrapKit/Controllers/TableViewControllers/BtSettingsAccountsTableViewController.swift
//
//  BtSettingsAccountsTableViewController.swift
//  bugTrap
//
//  Created by Colby L Williams on 6/30/14.
//  Copyright (c) 2014 bugTrap. All rights reserved.
//

import UIKit
import MessageUI

class BtSettingsAccountsTableViewController: BtTableViewController, MFMailComposeViewControllerDelegate {
    
	@IBOutlet var versionLabel: UILabel!
	
    var trackerTypes : [TrackerStatus : [TrackerType]]?
	var trackerUsers : [TrackerType : String]?
    
    override func viewDidLoad() {
		super.viewDidLoad()
		
		screenName = GAIKeys.ScreenNames.trackerAccounts
		
		versionLabel.text = Settings.versionBuildString
		
        trackerTypes = TrackerService.Shared.getTrackerTypesByStatus()
		
		trackerUsers = TrackerService.Shared.getUsernamesForEnabledTrackerTypes(trackerTypes![TrackerStatus.Enabled]!)
    }

	
	
	@IBAction func DoneButtonClick(sender: AnyObject) {
		navigationController!.dismissViewControllerAnimated(true, completion: nil)
	}
	

	
	@IBAction func letUsKnowClicked(sender: AnyObject) {
		if MFMailComposeViewController.canSendMail() {
			let mailComposeViewController = configuredMailComposeViewController()
			presentViewController(mailComposeViewController, animated: true, completion: nil)
		} else {
			let sendMailErrorController = UIAlertController(title: "Could Not Send Email", message: "Your device could not send e-mail.  Please check e-mail configuration and try again.", preferredStyle: .Alert)
			presentViewController(sendMailErrorController, animated: true, completion: nil)
		}
	}
	
	
	func configuredMailComposeViewController() -> MFMailComposeViewController {
		let mailComposerVC = MFMailComposeViewController()
			mailComposerVC.mailComposeDelegate = self
		
			mailComposerVC.setToRecipients([AppKeys.ContactEmail.string])
			mailComposerVC.setSubject("App: Tracker Integration")
			mailComposerVC.setMessageBody("Please provide the Tracker\'s name, a link to its website, and a link to the API documentation.", isHTML: false)
		
		return mailComposerVC
	}
	
	
	func mailComposeController(controller: MFMailComposeViewController, didFinishWithResult result: MFMailComposeResult, error: NSError?) {
		controller.dismissViewControllerAnimated(true, completion: nil)
	}
	
	
	// #pragma mark - Table view data source
	
	override func numberOfSectionsInTableView(tableView: UITableView?) -> Int {
		return 2
	}
	
	override func tableView(tableView: UITableView?, numberOfRowsInSection section: Int) -> Int {
		return section == 0 ? trackerTypes![TrackerStatus.Enabled]!.count : trackerTypes![TrackerStatus.ComingSoon]!.count
	}
		
	override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {

		let cell = tableView.dequeueReusableCellWithIdentifier("BtSettingsAccountTableViewCell") as! BtSettingsAccountTableViewCell
		
		let status = TrackerStatus(rawValue: indexPath.section + 2)!
		
        let tracker = trackerTypes![status]![indexPath.row]
		
		let enabled = indexPath.section == 0
		
        cell.textLabel?.text = tracker.rawValue
		cell.textLabel?.textColor = enabled ? Colors.Black.uiColor : Colors.Gray.uiColor
		
        cell.detailTextLabel?.text = enabled && tracker.enabled ? trackerUsers![tracker] : "coming soon"
        cell.detailTextLabel?.textColor = enabled ? Colors.Theme.uiColor : Colors.LightGray.uiColor
        cell.accessoryType = enabled ? .DisclosureIndicator : .None
		if let imageName = tracker.smallImageName {
			cell.imageView?.image = UIImage(named: imageName)
		} else {
			cell.imageView?.image = nil
		}
		
		return cell
	}
	

	override func tableView(tableView: UITableView, shouldHighlightRowAtIndexPath indexPath: NSIndexPath) -> Bool {
		return indexPath.section == 0
	}
	
	
	override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
		
		tableView.cellForRowAtIndexPath(indexPath)!.selected = false
		
		if indexPath.section == 0 {
            
            let status = TrackerStatus(rawValue: indexPath.section + 2)!
            
            let tracker = trackerTypes![status]![indexPath.row]
			
			if let loginControllerName = tracker.loginControllerName {
				if let loginController = storyboard?.instantiateViewControllerWithIdentifier(loginControllerName) {
					showViewController(loginController, sender: self)
				}
			}
		}
	}
	
	
	override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
		return indexPath.section == 0 ? 44.0 : 40.0
	}
	
	
	override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String {
		return section == 0 ? "Issue Trackers" : ""
	}

	
	override func prefersStatusBarHidden() -> Bool {
		return true
	}
}

File: iOS/Code/Xamarin/bugTrap/bugTrapKit/Controllers/TableViewControllers/BtSettingsAccountsTableViewController.cs


using System;

using Foundation;
using UIKit;
using MessageUI;
using System.Collections.Generic;

namespace bugTrapKit
{
	public partial class BtSettingsAccountsTableViewController : BtTableViewController, IMFMailComposeViewControllerDelegate
	{

		Dictionary<TrackerStatus, List<TrackerType>> trackerTypes;
		Dictionary<TrackerType, string> trackerUsers;


		public BtSettingsAccountsTableViewController (IntPtr handle) : base(handle)
		{
		}

		public override void ViewDidLoad ()
		{
			base.ViewDidLoad();

			ScreenName = GAIKeys.ScreenNames.TrackerAccounts;

			versionLabel.Text = Settings.VersionBuildString;

			trackerTypes = TrackerService.Shared.GetTrackerTypesByStatus();

			trackerUsers = TrackerService.Shared.GetUsernamesForEnabledTrackerTypes(trackerTypes?[TrackerStatus.Enabled]);
		}


		partial void DoneButtonClick (UIBarButtonItem sender)
		{
			DismissViewController(true, null);

		}

		partial void letUsKnowClicked (UIButton sender)
		{
			if (MFMailComposeViewController.CanSendMail) {
				var mailComposeViewController = configuredMailComposeViewController();
				PresentViewController(mailComposeViewController, true, null);
			} else {
				var sendMailErrorController = new UIAlertController { 
					Title = "Could Not Send Email", 
					Message = "Your device could not send e-mail.  Please check e-mail configuration and try again."
				};

				PresentViewController(sendMailErrorController, true, null);
			}
		}


		#region MFMailComposeViewController

		MFMailComposeViewController configuredMailComposeViewController ()
		{
			var mailComposerVC = new MFMailComposeViewController ();
			mailComposerVC.MailComposeDelegate = this;

			mailComposerVC.SetToRecipients(new [] { AppKeys.ContactEmail });
			mailComposerVC.SetSubject(AppKeys.Contact.IntegrationSubject);
			mailComposerVC.SetMessageBody(AppKeys.Contact.IntegrationBody, false);

			return mailComposerVC;
		}

		[Export("mailComposeController:didFinishWithResult:error:")]
		public void Finished (MFMailComposeViewController controller, MFMailComposeResult result, NSError error)
		{
			controller.DismissViewController(true, null);
		}

		#endregion


		#region TableViewSource

		public override nint NumberOfSections (UITableView tableView) => 2;

		public override nint RowsInSection (UITableView tableview, nint section)
		{
			return section == 0 ? trackerTypes?[TrackerStatus.Enabled]?.Count ?? 0 : trackerTypes?[TrackerStatus.ComingSoon]?.Count ?? 0;
		}


		public override UITableViewCell GetCell (UITableView tableView, NSIndexPath indexPath)
		{
			var cell = tableView.DequeueReusableCell(BtSettingsAccountTableViewCell.CellReuseId) as BtSettingsAccountTableViewCell;

			var status = (TrackerStatus)(indexPath.Section + 2);

			var tracker = trackerTypes[status][indexPath.Row];

			var enabled = indexPath.Section == 0;

			cell.TextLabel.Text = tracker.ToString();
			cell.TextLabel.TextColor = enabled ? Colors.Black : Colors.Gray;

			cell.DetailTextLabel.Text = enabled && tracker.Enabled() ? trackerUsers[tracker] ?? string.Empty : "coming soon";
			cell.DetailTextLabel.TextColor = enabled ? Colors.Theme : Colors.LightGray;

			cell.Accessory = enabled ? UITableViewCellAccessory.DisclosureIndicator : UITableViewCellAccessory.None;

			cell.ImageView.Image = string.IsNullOrEmpty(tracker.SmallImageName()) ? null : UIImage.FromBundle(tracker.SmallImageName());

			return cell;
		}


		public override bool ShouldHighlightRow (UITableView tableView, NSIndexPath rowIndexPath) => rowIndexPath.Section == 0;


		public override void RowSelected (UITableView tableView, NSIndexPath indexPath)
		{
			tableView.CellAt(indexPath).Selected = false;

			if (indexPath.Section == 0) {

				var status = (TrackerStatus)(indexPath.Section + 2);

				var tracker = trackerTypes[status][indexPath.Row];

				if (!string.IsNullOrEmpty(tracker.LoginControllerName())) {
				
					var loginController = Storyboard.Instantiate(tracker.LoginControllerName());

					ShowViewController(loginController, this);
				}
			}
		}


		public override nfloat GetHeightForRow (UITableView tableView, NSIndexPath indexPath)
		{
			return indexPath.Section == 0 ? 44 : 40;
		}


		public override string TitleForHeader (UITableView tableView, nint section)
		{
			return section == 0 ? "Issue Trackers" : string.Empty;
		}

		#endregion


		public override bool PrefersStatusBarHidden () => true;

		public override UIStatusBarStyle PreferredStatusBarStyle () => UIStatusBarStyle.LightContent;
	}
}